### PR TITLE
New Visual Studio project generation logic that supports any platform that wants to opt in

### PR DESCRIPTION
**Pull request by shana**
Custom Visual Studio project generation logic that supports any platform via an opt-in mechanism, so Visual Studio can be used to run scons for any platform, with the right defines and file lists per target.

For a platform to opt-in to this, it needs to add a `msvs.py` script with platform-specific settings. This PR adds it for Windows, both as an example and to keep the same existing behaviour of the current vs proj generation.

Additionally, in order for the list of platform files on the Visual Studio project to be accurate, platforms should use `env.add_source_files()` and `env.platform_sources` like the rest of the codebase uses `env.core_sources` etc, filling out an array of obj files that can be parsed in the same way as all the other lists of sources are parsed. `env.platform_sources` is already set in the shared platform SCsub, this makes Windows use it as well. Other platforms should be fixed as well, but I didn't want to add more stuff to this PR, that can be done later.

### Basic usage

```
scons vsproj=yes
```

This will generate a solution and a vcxproj file containing configurations for all the platforms that are currently enabled for VS support (i.e. all platforms with a msvs.py file). Additionally, a `[projectname]_[platform]_[target]_[arch].generated.props` file is generated, containing the actual build definitions, file list and command line arguments to do a build for that specific configuration (for example, by default on Windows, it would generate `godot_windows_editor_x86_64.generated.props`)

To generate build configuration files for all platforms+targets+arch combinations that a user is interested in, they can call

  ```
scons vsproj=yes [additional required build flags]
```

for each combination of platform+target+arch. This will generate the relevant vs project files, but skip the build process.

This lets project files be quickly generated even if there are build errors. Once project files are generated for the desired platform+target combinations, builds can be done as normal via VS or the command line. Any valid platform+target+arch combination that wasn't generated will still show up in VS as an option, but won't do anything.

*New flag*

- `vsproj_gen_only=[yes|no] (default: yes)`: Controls whether the build is skipped or not.

When passing `vsproj=yes` to a build, it will generate the project files corresponding to that particular build configuration, but by default it won't trigger the actual command line build.

This is because if you're generating VS project files, it's probably because you want to build from Visual Studio, and whatever build you run on the command line will likely be thrown away when you build from VS. scons detects slight changes in the environment between the command line and VS, and decides it needs to do a new build even if all the build command arguments are the same.

If you really want to generate the project files and build at the same time from the command line, `vsproj_gen_only=no`

### What this fixes

Currently, VS project generation is hardcoded to the windows platform, which means anyone wanting to use Visual Studio as an editor for other platform targets that can be cross-compiled from Windows is out of luck.

Additionally, the configurations created for the non-active target at the time of generation are wrong - scons only knows the DEFINES/FLAGS/etc specific to the current platform+target+arch, which means any of the other targets are going to have the wrong configuration data set on the vcxproj. For example, if the vsproj flag is set to yes on an editor target, the template_debug and template_release vcxproj configurations will end up with the editor defines and settings. The `scons` command will be correct for each target, but the defines and includes will be wrong, so Visual Studio will show the wrong information in the editor. This is not ideal.

**The only way to properly generate vcxproj Property blocks that have the correct defines/includes/flags/etc for a platform+target+arch combination, is to call scons separately and have it generate the vcxproj section for that active target, once per combo.** That is what this PR enables.

To enable VS project generation support for a platform, platforms can add a `msvs.py` file that contains the list of supported architectures, the list of supported Godot targets, and an optional build prefix to use before invoking scons, usually a batch file that sets up the environment for the command line build.

Platforms with a `msvs.py` file will be added to the solution and project, but they will not be buildable from Visual Studio until the specific props file for that platform+target+arch combination is generated. This is because we can only know what the right defines/includes/flags/etc are on the active build target currently being processed by scons, so each combo needs to be generated by a separate scons call. All combos can be generated very quickly with the command line noted above. If there are missing props files, the project configuration will still be in the solution, it just won't do anything when built.

Platforms that don't support an editor target will have a dummy editor target that won't do anything on build, but will have the files and configuration for the windows editor target.

PS: I'm currently successfully using this to build for Windows editor+template/debug+release/x86+x64 and Switch/GDK Windows/Xbox/PS5 template/debug+release/arches. and it's making me very happy.

_Contributed with ❤️ by W4Games_

### CODE REVIEW

**Comment by rxlecky** [comment]
Regarding `vsproj_gen_only` option, I'd honestly say that it should be the default behaviour. Like, anytime a user is compiling manually from a command line with `vsproj=yes`, I think they never actually want to build, they just want to generate the project files and then build from Visual Studio. The reason is that right now since there's no option to only generate the project files, the first time that the user downloads the source and builds, it builds the engine the first time from command line with `scons vsproj=yes` and then when launching the project from VisualStudio, the engine is built the second time, even if it just finished building - presumably because the Visual Studio build command is using slightly different SCons arguments than when building manually from command line and hence SCons thinks it needs to do a full rebuild.

**Comment by shana** [comment]
> Regarding `vsproj_gen_only` option, I'd honestly say that it should be the default behaviour. Like, anytime a user is compiling manually from a command line with `vsproj=yes`, I think they never actually want to build, they just want to generate the project files and then build from Visual Studio. The reason is that right now since there's no option to only generate the project files, the first time that the user downloads the source and builds, it builds the engine the first time from command line with `scons vsproj=yes` and then when launching the project from VisualStudio, the engine is built the second time, even if it just finished building - presumably because the Visual Studio build command is using slightly different SCons arguments than when building manually from command line and hence SCons thinks it needs to do a full rebuild.

The thing with the original vsproj generation is that it happens after the build, not before, which means this option wouldn't work there. That's one of the reasons why I made this new generation code, is so that we can generate project files without going through the build process (because yes, having to wait for the build in order to have project files that I can use to figure out why the build is failing is, erm, not ideal).

Also yeah, the reason it currently builds twice is because the command line is oh so very slightly different, enough that scons considers it a separate build. That is hopefully not the case anymore with this new project generation, even if you do let it build from the command line first (but now you can skip that bit, so yay)

**Comment by shana** [comment]
I've been super busy closing down other work and couldn't finish this up, but I'm going to get back to it before the end of this week. I've been testing it, and I just have a couple of small bugfixes to add to this, rebase, and it should be ready to go.

**Comment by rxlecky** [comment]
> The thing with the original vsproj generation is that it happens after the build, not before, which means this option wouldn't work there. That's one of the reasons why I made this new generation code ...

I wasn't questioning your solution by any means, I understand there are limitations with the current VS project generation code that need to be overcome. I was more so commenting on the naming and structure of the build system parameters. Currently, I feel Godot build system is at the point where there are a lot of basic parameters and switches that allow a good level of control over the build process, without being too overwhelming. There are high-level switches that allow setting multiple lower-level parameters for the most common use case and I think this is a good approach. My suggestion is that we should have a similar high-level switch for just generating the VS project files without building, as I find that's the most common use case. It's something that wasn't possible until now due to the shortcomings in the project generation code that you now solved and I think it should become the new default.

So my suggestion is that we should have a shorthand for `scons vsproj=yes use_vsproj2=yes vsproj_gen_only=yes` as that operation should really be the default of what the build system does when the user requests generating VS project files. Either something like `vsproj_gen=yes` or even the `vsproj=yes` could be repurposed if there aren't a lot of dependencies on this build + generate behaviour of the current `vsproj=yes` switch. The only dependency I found within Godot itself is that the generated VS rebuild command is using this behaviour to achieve project generation and a clean rebuild, but I don't see a reason why this couldn't easily be changed into two scons calls, the first one doing the project generation and the other one doing the build, especially since the project generation should be really quick now with your changes.

Anyway, this may not be the right place to discuss this and this kind of change would be better suited as a follow-up PR after your changes are merged and after a thorough discussion. Thanks for the great work on this! ❤️

**Comment by shana** [comment]
Well, Christmas happened so there went my week. Gonna fix the conflicts and push add the final bugfix for this one.

> So my suggestion is that we should have a shorthand for scons vsproj=yes use_vsproj2=yes vsproj_gen_only=yes as that operation should really be the default of what the build system does when the user requests generating VS project files

Ah, gotcha!

Ok, so, I didn't want to change the `vsproj` option to use this new generation logic by default so it wouldn't be a surprise to users that might be using the existing one.

Also, the current logic always generates and builds (because it has to build before it generates, the build step is always required), so I kept that in v2 - again, because of existing user expectations. But yes, if I'm doing vsproj generation, most of the times I only want to generate on the CLI. If there's not a lot of users depending on this feature, perhaps it's not worth keeping the old defaults and carrying that UX debt.

So my question to maintainers is:
- Should I make `vsproj=yes` use this new v2 logic by default?
- Should I make this generate-only by default?

/cc @akien-mga

**Comment by shana** [comment]
I've done the final tweaks, and this is conflict-free and ready to go. I'm going to keep the current defaults of generating and building for now, because making generate-only the default requires changing the flags I set in the VS rebuild commands, and I would like to test that more, but separately.

We're already taking a dependency on this branch for the W4 consoles fork, so I would really like to get this merged in its current state with no additional changes or rebases, because a force-push on this branch will cause very annoying merge conflicts later on when this gets merged and we pull in godot master into our fork, and I really don't need more merges in my life :P

**Comment by adamscott** [comment]
> * Should I make `vsproj=yes` use this new v2 logic by default?

I don't use personally vsproj files, but IMHO, yes. The current way of generating vsproj is flawed and should be replaced with this PR. But we should at least provide a migration guide.

**Comment by shana** [comment]
Ok, so during more testing we found a bug, so might as well do more polish on this branch, and I'll take the merge conflict hit later.

Since I'm touching this again, I've removed the old vs generation logic entirely and made this one the default.

I've also gone ahead and made it not build by default if the `vsproj` flag is set, because generating all the combinations that you want to use in VS up front, and then build from VS is the most likely user scenario. `vsproj_gen_only=no` is available for making it generate AND build. Should I add this flag to SConstruct?

I've added some extra logic so that platforms can mark files for inclusion that won't be autodetected, and made the parsing of the file list more robust (so platforms can add files as strings, or File() objects, etc).

File: .gitignore
Comment (AThousandShips) [inline .gitignore:370]:
```suggestion
*.generated.props

```

**Comment by shana** [comment]
Rebased to pick up unrelated CI fixes and squash everything.

File: misc/msvs/props.template
Comment (akien-mga) [inline misc/msvs/props.template:21]: Missing newline at EOF for all the XML templates.

**Review by akien-mga — APPROVED** [review]
Looks great! Didn't review the new logic in depth, but the reasoning makes a lot of sense, and the added flexibility will be good to have.

And I'm aware of a lot of problems with the old VS proj generation so I think restarting from a clean implementation is a good idea.

File: misc/msvs/props.template
Comment (akien-mga) [reply misc/msvs/props.template:21]: If the line endings are `\r\n`, then the EOF would not be needed IMO. We have CI checks to ensure that all files use Unix like format, but somehow it doesn't seem to be triggered here. Either way if we need those to be kept in Windows-like format we need to make sure it's properly enforced by `.gitattributes` and not overruled by our CI scripts.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
